AirCasting/ABConnector/BluetoothManager.swift

```diff
@@ -68,6 +73,48 @@ class BluetoothManager: NSObject, ObservableObject {
             let _ = centralManager
         }
     }
+
+    // MARK: - Refactored part
+    // This is the part of this class that is already refactored.
+
+    typealias CharacteristicObserverAction = (Result<Data?, Error>) -> Void
+
+    private struct CharacteristicObserver {
+        let identifier = UUID()
+        let action: CharacteristicObserverAction
+    }
+
+    // The mapping is CBUUID -> CharacteristicObserver
+    // and not UUID(observation token) -> CharacteristicObserver
+    // because we care more about the update performance than un-register performance.
+    private var charactieristicsMapping: [CBUUID: [CharacteristicObserver]] = [:]
+    private let characteristicsMappingLock = NSRecursiveLock()
+
+    /// Adds an entry to observers of a particular characteristic
+    /// - Parameters:
+    ///   - characteristic: UUID of characteristic to observe
+    ///   - notify: block called each time characteristic changes (either changes value or throws an error)
+    /// - Returns: Opaque token to use when un-registering
+    func subscribeToCharacteristic(_ characteristic: CBUUID, notify: @escaping CharacteristicObserverAction) -> AnyHashable {
+        let observer = CharacteristicObserver(action: notify)
+        characteristicsMappingLock.lock()
```

Comment:
I understand that the NSRecursiveLock is used here to make this function safe to use by multiple threads at the same time? I'm curious why this was important with this method?

Reply:
Yes. This is a pretty low-level class, so we need to make it thread-safe in order to support multiple clients calling those methods from multiple threads. For characteristics observing we're using a dictionary mapping so our critical section will be reads and writes to it. `NSLock` wouldn't work here, because we need to allow to unsubscribe/subscribe from the `notify` block which is invoked inside a locked part currently (deadlock scenario `NSRecursiveLock` is solving). 
We want to move those `notify` calls to yet a different thread in order to ensure time efficiency - when we implement this we'd be able to switch back to the `NSLock`. I know this is pretty low-level and usually thread-safety is done using GCD, but in this case a simple mutex is sufficient and has a huge benefit of speed (it's waaay faster than usual scheduling to serial queues), and bluetooth is one of those few places that we really need efficiency at